contracts/utils/structs/CircularBuffer.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Math} from "../math/Math.sol";
+import {Arrays} from "../Arrays.sol";
+import {Panic} from "../Panic.sol";
+
+/**
+ * @dev A buffer of items of fixed size. When a new item is pushed, it takes the place of the oldest one in the buffer
+ * so that at all times, only the last N elements are kept. Items cannot be removed. The entier buffer can be reset.
+ * Last N elements can be accessed using their index from the end.
+ *
+ * Complexity:
+ * - insertion (`push`): O(1)
+ * - lookup (`last`): O(1)
+ * - inclusion (`includes`): O(N) (worst case)
+ * - reset (`clear`): O(1)
+ *
+ * * The struct is called `Bytes32CircularBuffer`. Other types can be cast to and from `bytes32`. This data structure
+ * can only be used in storage, and not in memory.
+ * ```solidity
+ * CircularBuffer.Bytes32CircularBuffer buffer;
+ * ```
+ */
+library CircularBuffer {
+    /**
+     * @dev Counts the number of items that have been pushed to the buffer. The residu modulo _data.length indicates
+     * where the next value should be stored.
+     *
+     * Struct members have an underscore prefix indicating that they are "private" and should not be read or written to
+     * directly. Use the functions provided below instead. Modifying the struct manually may violate assumptions and
+     * lead to unexpected behavior.
+     *
+     * The last item is at data[(index - 1) % data.length] and the last item is at data[index % data.length]. This
+     * range can wrap around.
+     */
+    struct Bytes32CircularBuffer {
+        uint256 _count;
+        bytes32[] _data;
+    }
+
+    /**
+     * @dev Initialize a new CircularBuffer of given length.
+     *
+     * If the CircularBuffer was already setup and used, calling that function again will reset it to a blank state.
+     */
+    function setup(Bytes32CircularBuffer storage self, uint256 length) internal {
+        clear(self);
+        Arrays.unsafeSetLength(self._data, length);
+    }
+
+    /**
+     * @dev Clear all data in the buffer, keeping the existing length.
+     */
+    function clear(Bytes32CircularBuffer storage self) internal {
+        self._count = 0;
+    }
```

Comment:
While this does not clear the storage, it reset the buffer to that, through the API, it behaves just like if it never contained any data.

Anything store in the array is ignored because count is 0, and whenever the count increases, the new values replace the old ones.

Reply:
Yes, all good if the memory is only used through the API, but more experienced users may unexpectedly find dirty memory when they thought `clear` actually clears the data as the NatSpec suggests.

I would be better if we are a bit more specific

```suggestion
    /**
     * @dev Clear all data in the buffer without resetting memory, keeping the existing length.
     */
    function clear(Bytes32CircularBuffer storage self) internal {
        self._count = 0;
    }
```